documents/JupyterHub.md

```diff
@@ -0,0 +1,104 @@
+spawner" but its resoruce limiting is known to be quirky.
+(See [this talk](https://research.cs.wisc.edu/htcondor/HTCondorWeek2017/presentations/WedDownes_cgroups.pdf).)
+
+## Data Access
+
+### User Interface
+
+Users will access all data through databroker. Currently, that means using
+databroker's Python API to load the data into a Python process and either work
+on it their or export it to files. In time, an HTTP API, a web interface, and
+other GUI interfaces will be developed by DAMA and other bluesky collaborators.
+
+### Databases
+
+We will make the central MongoDB instance visible from SDCC. (???)
+
+### Files from Detectors
+
+We will continually sync the data from beamlines to storage at SDCC, so that any
+outages on the NSLS-II side only affect the availability of *new* data, not
+*all* saved data. The data will move from the beamline storage through the
+NSLS-II central GPFS to the NSLS-II's data transfer node, to SDCC's data
+transfer node, to ???
+
+## Kernel Management 
+
+### Status Quo
+
+Currently, DAMA deploys one Juptyer kernel and associated conda environment per
+beamline, per cycle---sometimes a couple per cycle, for minor updates. Because
+we provide long-term support for these, we have accumulated a large number of
+kernels and many environments that are very similar but not identical.
+
+### Long-Term Support Kernels
+
+At the start of a cycle, deploy a small set of "stable" kernels+environments
+that will receive long-term support. Instead of making one per beamline,
+start with one "NSLS-II base" environment with the union of all requirements
+(csxtools, sixtools, etc.) In the future, create additional environments as
+needed when we encounter conflicting or significantly different set of
+requirements.  This is similar Jupyter's
+[docker stacks](https://github.com/jupyter/docker-stacks), with images like
+"minimal", "scipy", "data science", "pyspark", etc.
+
+### Fast Delivery of New Code via "Latest" Kernel
+
+Every night, automatically build a new environment with a unique name. This will
+pick up the latest versions of all the packages in that environment, such as
+sixtools. For situations where "nightly" is not fast enough, where staff want to
+deliver a new update immediately, provide a way for staff to trigger the build
+manually at any time (say, with a button on a webpage).
+
+Provide *one* kernel, "NSLS-II base latest", whose target environment is
+automatically moved to point to the newest environment every time a build runs.
+This means that every time a notebook on the "NSLS-II base latest" has its
+kernel restarted, it may be pointing at a different environment than before.
+Automatically delete old "latest" environments, but not *immediately*. Notebooks
+that have been running for a couple days will still be attached to the
+environment that was newest *when they started* and that environment should not
+be ripped out from underneath them while they are still running.
```

Comment:
We may just want to tag it as `YYYYmmDD:HH:MM`, this way there will be no confusion what is the _latest_ kernel. I tried to follow the procedure here before I discovered the CI building feature based on the tagged versions: https://hub.docker.com/r/nsls2/debian-with-miniconda/tags.

Reply:
Yes. Per discussion with @tacaswell this morning, we should also stash a YAML file from `conda env export ...` with that timestamp in the file name in a directory and keep it forever. That way, even as long "latest" environments are culled, it will always be possible to recreate an old one or go back through and do "archaeology" to sort out which environment would have been active when.